### PR TITLE
Unbreak crowbar_batch with rails4 port (bnc#912309)

### PR DESCRIPTION
**Pull request by vuntz**
Looks like the replies are different now...

http://bugzilla.suse.com/show_bug.cgi?id=912309

### CODE REVIEW

File: bin/crowbar_batch [hunk 1]
```diff
@@ -458,7 +458,7 @@ def export_barclamp(barclamp, alias_by_host)
     end
   end
 
-  proposals = get_proposal_list(barclamp)
+  proposals = get_proposal_list(barclamp)[barclamp]
```
Comment (tboerger) [inline bin/crowbar_batch]: That feels wrong for me

Comment (vuntz) [inline bin/crowbar_batch]: Why? Try it :-)

Comment (tboerger) [inline bin/crowbar_batch]: Yeah, but calling it with barclamp and getting a barclamp entry afterwards?

**Comment by vuntz** [comment]
FWIW, I now hit this:
```
crowbar batch build foo
[19:54:44] crowbar barclamp, 'default' proposal:
   Already exists
/usr/lib64/ruby/gems/2.1.0/gems/chef-10.32.2/lib/chef/mixin/deep_merge.rb:207:in `deep_merge!': Use RbConfig instead of obsolete and deprecated Config.
/usr/lib64/ruby/gems/2.1.0/gems/chef-10.32.2/lib/chef/mixin/deep_merge.rb:207: warning: toplevel constant Config referenced by Chef::Config
/usr/lib64/ruby/gems/2.1.0/gems/chef-10.32.2/lib/chef/mixin/deep_merge.rb:207:in `deep_merge!': Use RbConfig instead of obsolete and deprecated Config.
/opt/dell/bin/barclamp_lib.rb:536:in `eval': undefined method `[]' for RbConfig:Module (NoMethodError)
        from /usr/lib64/ruby/gems/2.1.0/gems/chef-10.32.2/lib/chef/mixin/deep_merge.rb:207:in `deep_merge!'
        from /usr/lib64/ruby/gems/2.1.0/gems/chef-10.32.2/lib/chef/mixin/deep_merge.rb:138:in `block in deep_merge!'
        from /usr/lib64/ruby/gems/2.1.0/gems/chef-10.32.2/lib/chef/mixin/deep_merge.rb:133:in `each'
        from /usr/lib64/ruby/gems/2.1.0/gems/chef-10.32.2/lib/chef/mixin/deep_merge.rb:133:in `deep_merge!'
        from /usr/lib64/ruby/gems/2.1.0/gems/chef-10.32.2/lib/chef/mixin/deep_merge.rb:138:in `block in deep_merge!'
        from /usr/lib64/ruby/gems/2.1.0/gems/chef-10.32.2/lib/chef/mixin/deep_merge.rb:133:in `each'
        from /usr/lib64/ruby/gems/2.1.0/gems/chef-10.32.2/lib/chef/mixin/deep_merge.rb:133:in `deep_merge!'
        from /usr/lib64/ruby/gems/2.1.0/gems/chef-10.32.2/lib/chef/mixin/deep_merge.rb:138:in `block in deep_merge!'
        from /usr/lib64/ruby/gems/2.1.0/gems/chef-10.32.2/lib/chef/mixin/deep_merge.rb:133:in `each'
        from /usr/lib64/ruby/gems/2.1.0/gems/chef-10.32.2/lib/chef/mixin/deep_merge.rb:133:in `deep_merge!'
        from /usr/lib64/ruby/gems/2.1.0/gems/chef-10.32.2/lib/chef/mixin/deep_merge.rb:138:in `block in deep_merge!'
        from /usr/lib64/ruby/gems/2.1.0/gems/chef-10.32.2/lib/chef/mixin/deep_merge.rb:133:in `each'
        from /usr/lib64/ruby/gems/2.1.0/gems/chef-10.32.2/lib/chef/mixin/deep_merge.rb:133:in `deep_merge!'
        from ./crowbar_batch:330:in `merge_attributes'
        from ./crowbar_batch:249:in `modify_proposal'
        from ./crowbar_batch:182:in `build_proposal'
        from ./crowbar_batch:112:in `block in build'
        from ./crowbar_batch:111:in `each'
        from ./crowbar_batch:111:in `build'
        from (eval):1:in `run_sub_command'
        from /opt/dell/bin/barclamp_lib.rb:536:in `eval'
        from /opt/dell/bin/barclamp_lib.rb:536:in `run_sub_command'
        from /opt/dell/bin/barclamp_lib.rb:540:in `run_command'
        from ./crowbar_batch:553:in `main'
        from ./crowbar_batch:556:in `<main>'
```

File: bin/crowbar_batch (hunk 1 above)
Comment (vuntz) [inline bin/crowbar_batch]: I'm not the one who changed the API ;-) What is returned now is something like `["crowbar", ["default"]]`

**Comment by vuntz** [comment]
Ok, seems to work now.

File: bin/crowbar_batch (hunk 1 above)
Comment (tboerger) [inline bin/crowbar_batch]: Yeah but this looks like an unwanted change of the API. Currently without accessing my codebase i have no idea where this have been introduced.

**Comment by tboerger** [comment]
IMHO this is the wrong workaround. It looks like an unwanted API change.

**Comment by bkutil** [comment]
So what seems to have happened here was, that the routes got changed (mea maxima culpa as I was responsible for fixing them after the transition), and then we observed the issue with our tools failing and fixed them.

So now we have two possible approaches:

1) taken by this patch is to fix the tools and keep the API as it is. As long as we're sure all our tools now work and that customers don't need cloud4 vs cloud5 compatibility between tools and the API, we're OK.
2) I go and fix the API, i.e. write a test that does the same request as the crowbar batch here and make sure that the code behaves the same under rails 2 and rails 4. Then this PR would be not needed and we're done (unless we'll have to revert some other PRs that were made to fix this in the meantime)

@vuntz @tboerger what do you think about this?

**Comment by vuntz** [comment]
@bkutil I don't have any opinion on what's best, tbh.

FWIW, you can also see the API change by doing `crowbar $barclamp list` and other similar commands. So it's not just `crowbar batch`.

**Comment by tboerger** [comment]
I prefer to fix the API

**Comment by aspiers** [comment]
Right - `crowbar batch` doesn't contain any API code itself, it just reuses the same code that all the other CLI commands use.  So whatever approach you decide, please check they all still work ;-)

**Comment by bkutil** [comment]
@tboerger the API / routes are unchanged, however, rails renders JSON with a root. A simple fix might be just to pas `:root => false` to the `render` calls, but there might be a global option (I was unable to find it after googling for some time, and ActiveRecord::Base.include_root_in_json does not seem to have an effect here).

**Comment by tboerger** [comment]
It must be cause by something different. https://github.com/crowbar/barclamp-crowbar/blob/master/crowbar_framework/config/initializers/wrap.rb

**Comment by bkutil** [comment]
@tboerger you can check for yourself, if you're willing to spend 10 minutes on it - copy the line from inside the block setting the activerecord::base option before the render in e.g. index action in barclamps controller, restart, then curl that endpoint. The setting will have no effect (at least that is what I observed). Maybe because what we're trying to render is not an activerecord base object?

**Comment by vuntz** [comment]
Moved to https://github.com/crowbar/barclamp-crowbar/issues/1181 and https://github.com/crowbar/barclamp-crowbar/issues/1185

**Pull request closed without merging**